routes/s3handler.js

```diff
@@ -0,0 +1,239 @@
+import express from 'express'
+import CryptoJS from 'crypto-js'
+import bodyParser from 'body-parser'
+import getObjectURLFromS3 from '../src/utils/s3'
+import faceUtils from '../src/utils/face++'
+
+const clientSecretKey = process.env.AWS_SECRET_ACCESS_KEY
+const expectedMinSize = 0
+const expectedMaxSize = 5000000
+const expectedBucket = 'uploadedphotostomatch'
+const expectedHostname = 'uploadedphotostomatch.s3.amazonaws.com'
+
+const router = express.Router()
+
+// Signs any requests.  Delegate to a more specific signer based on type of request.
+function signRequest(req, res) {
+    if (req.body.headers) {
+        signRestRequest(req, res);
+    }
+    else {
+        signPolicy(req, res);
+    }
+}
+
+// Signs multipart (chunked) requests.
+function signRestRequest(req, res) {
+    var version = req.query.v4 ? 4 : 2,
+        stringToSign = req.body.headers,
+        signature = version === 4 ? signV4RestRequest(stringToSign) : signV2RestRequest(stringToSign);
+
+    var jsonResponse = {
+        signature: signature
+    };
+
+    res.setHeader("Content-Type", "application/json");
+
+    if (isValidRestRequest(stringToSign, version)) {
+        res.end(JSON.stringify(jsonResponse));
+    }
+    else {
+        res.status(400);
+        res.end(JSON.stringify({invalid: true}));
+    }
+}
+
+function signV2RestRequest(headersStr) {
+    return getV2SignatureKey(clientSecretKey, headersStr);
+}
+
+function signV4RestRequest(headersStr) {
+    var matches = /.+\n.+\n(\d+)\/(.+)\/s3\/aws4_request\n([\s\S]+)/.exec(headersStr),
+        hashedCanonicalRequest = CryptoJS.SHA256(matches[3]),
+        stringToSign = headersStr.replace(/(.+s3\/aws4_request\n)[\s\S]+/, '$1' + hashedCanonicalRequest);
```

Comment:
Is there a better way to do this than parsing the header with Regex? Seems kind of fragile if something changes and it doesn't find a match; might propagate error downstream.